### DISCUSSION TITLE
Is it possible to remove a reported metric?

### DISCUSSION THREAD

**Discussion by tomkerkhove** [PrometheusClientNet/Prometheus.Client#177] [Q&A] [answered]
Is it possible to remove a reported metric?

I'm afraid the answer will be no, but you never know.

Relates to https://github.com/tomkerkhove/promitor/issues/1923

**Comment by phnx47** [comment]
I will investigate about removing a reported metric

About removing 'Labels', we Should implement this:
> Metrics with labels SHOULD support a remove() method with the same signature as labels() that will remove a Child from the metric no longer exporting it, and a clear() method that removes all Children from the metric. These invalidate caching of Children.

Just to remove labels, is it enough?

**Reply by tomkerkhove** [reply]
That would be useful as well but we actually need to remove a full metric in this case.

**Comment by sanych-sun** [comment, marked as answer]
You can use [Remove](https://github.com/prom-client-net/prom-client/blob/13f06ec9ea27d37992cae668d78559fd7efce7dd/src/Prometheus.Client.Abstractions/Collectors/ICollectorRegistry.cs#L18) method of the ICollectorRegistry to remove all samples of the metric:

`registry.Remove("my-metric")`

Or [RemoveLabelled](https://github.com/prom-client-net/prom-client/blob/13f06ec9ea27d37992cae668d78559fd7efce7dd/src/Prometheus.Client.Abstractions/IMetricFamily.cs#L28) on the metric family to remove the single sample.

**Reply by phnx47** [reply]
Perfect :) I forgot to check `IMetricFamily` 

I'm not sure that using `Remove()` from `ICollectorRegistry` is ok. I will do some experiments with this.

**Reply by sanych-sun** [reply]
I believe it should be OK to use. As it simply removes the collector (metric family) from the registry. So it will be lost by metric factory and scrape endpoint

**Reply by phnx47** [reply]
@sanych-sun I mean, that usually you are working with `IMetricFactory`, not with `ICollectorRegistry`. I think about passing some methods from `ICollectorRegistry` to `IMetricFactory`.

**Reply by tomkerkhove** [reply]
That is correct, I only use `IMetricFactory`

**Reply by sanych-sun** [reply]
Good idea, we probably can add Release method(s), that accepts MetricFamily and remove it from the underlying registry. I've create [appropriate issue](https://github.com/prom-client-net/prom-client/issues/178). I will keep you posted and let you know once the changes will be done. 

@tomkerkhove meanwhile you can use the approach with access to ICollectorRegistry if it's an urgent requirement for your project.

**Reply by tomkerkhove** [reply]
Thanks a ton!

**Reply by phnx47** [reply]
Implemented #178, Published [v5.0.0](https://github.com/prom-client-net/prom-client/releases/tag/v5.0.0). I will mark as Answered!

**Reply by tomkerkhove** [reply]
Awesome, thank you!
